### PR TITLE
Rewritten RX/TX routines for vmxnet3 driver

### PR DESCRIPTION
**Pull request by coreybonnell-sophos**
vmxnet3 driver implementation has been largely modified such that the driver no longer memcpy's RX/TX buffer contents into netmap buffers, but rather configures the virtual device on initialization to point directly to the netmap buffers (such that they are populated directly via DMA). By doing so, a significant performance benefit was realized (5-10% increase for RX, 2x+ for TX). In addition, these changes take into account recent revisions to the vmxnet3 vNIC so that newer revisions are now supported (thus addressing https://github.com/luigirizzo/netmap/issues/575).

We have done our testing against kernel version 4.13. We have not tested previous versions, especially considering that the earliest supported kernel version for the current vmxnet3 driver is 3.10, which has been EOL'ed for over a year. As such, we created a separate header file for this new version of the driver.

Header file source code has already been formatted using clang-format (as was requested when the initial vmxnet3 driver was submitted as a PR).

### CODE REVIEW

**Comment by vmaffione** [comment]
Thanks! As a first step, I think you should rename `LINUX/final-patches/vanilla--vmxnet3--31000--99999` to `LINUX/final-patches/vanilla--vmxnet3--31000--40d00`, so that the old patch applies from 3.10 to 4.12 and the new one from 4.13 onwards. Then we can let travis build on all the versions and see what happens.

**Comment by coreybonnell-sophos** [comment]
Thanks @vmaffione, I just renamed the "old" patch file such that it will be applied on older kernels only as you requested (we had it renamed internally but I failed to include that rename in this PR).

**Comment by giuseppelettieri** [comment]
Actuall both patches should be removed from the PR, since they will be generated from the other repository (https://github.com/giuseppelettieri/linux-pspat). Sorry for the confusion.

File: LINUX/if_vmxnet3_netmap_v2.h [hunk 1]
```diff
@@ -0,0 +1,555 @@
...
+				vmxnet3_getRxDesc(
+				        rxd,
+				        &cmd_ring->base[cmd_ring->next2fill]
+				                 .rxd,
+				        &rxCmdDesc);
+				BUG_ON(!rxd->addr);
+
+				/* Recv desc is ready to be used by the device
+				 */
+				rxd->gen = cmd_ring->gen;
```
[the remark this answers is not in the document]
Comment (coreybonnell-sophos) [reply LINUX/if_vmxnet3_netmap_v2.h]: Hmm, good point about shared ownership of the buffer between userspace and the NIC. I'm wondering now if we should be resetting the RX descriptor gen in the "second part (skip past packets that userspace has released)" to ensure a clean hand-off of the buffer from userspace back to the NIC.

I'm unsure if the NIC will actually skip descriptors. Unfortunately, there is no datasheet available for vmxnet3, so I had to read various OS's implementations (such as BSD, Linux, etc.) of the vmxnet3 driver to determine how the vNIC operates. They all had comments indicating that the device can skip descriptors, so I added that here as well (I realize this is an unsatisfying response, but I couldn't find anything more definitive).

I'll create a new PR with the just the header file as you requested.

**Pull request closed without merging**